<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<title>hapFLK: a gentle introduction</title>
<meta name="author" content="(Bertrand Servin)"/>
<style type="text/css">
.underline { text-decoration: underline; }
</style>
<link rel="stylesheet" href="http://cdn.jsdelivr.net/reveal.js/3.0.0/css/reveal.css"/>

<link rel="stylesheet" href="http://cdn.jsdelivr.net/reveal.js/3.0.0/css/theme/beige.css" id="theme"/>

<link rel="stylesheet" href="mystyle.css"/>
<link rel="stylesheet" href="http://cdn.jsdelivr.net/reveal.js/3.0.0/lib/css/zenburn.css"/>
<!-- If the query includes 'print-pdf', include the PDF print sheet -->
<script>
    if( window.location.search.match( /print-pdf/gi ) ) {
        var link = document.createElement( 'link' );
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = 'http://cdn.jsdelivr.net/reveal.js/3.0.0/css/print/pdf.css';
        document.getElementsByTagName( 'head' )[0].appendChild( link );
    }
</script>
<script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
</head>
<body>
<div class="reveal">
<div class="slides">
<section id="sec-title-slide"><h1 class="title">hapFLK: a gentle introduction</h1><h2 class="author">Bertrand Servin</h2><h2 class="email"><a href="mailto:bertrand.servin@inra.fr">bertrand.servin@inra.fr</a></h2><h2 class="date">SSMPG Sept. 11-15 2017, Aussois, France</h2>
</section>

<section>
<section id="slide-orge00b072">
<h2 id="orge00b072">Resources</h2>
<ul>
<li>You can access this document as an HTML presentation <a href="http://htmlpreview.github.io/?https://github.com/BertrandServin/SSMPG2017/blob/master/hapflk.html">here</a>.</li>
<li>The hapFLK webpage is <a href="https://forge-dga.jouy.inra.fr/projects/hapflk">here</a></li>
<li>The hapflk software is available on <a href="https://pypi.python.org/pypi/hapflk">Pypi</a>.</li>
<li>Associated papers are:
<ul>
<li><a href="http://www.genetics.org/content/193/3/929.abstract">Fariello et al., 2013, Detecting Signatures of Selection Through Haplotype Differentiation Among Hierarchically Structured Populations. Genetics 193(3):929-941.</a></li>
<li><a href="http://www.genetics.org/content/186/1/241.abstract">Bonhomme et al., 2010, Detecting selection in population trees:
The Lewontin and Krakauer test extended. Genetics 186(1) 241-262</a></li>

</ul></li>

</ul>


</section>
</section>
<section>
<section id="slide-orgfbb9e65">
<h2 id="orgfbb9e65">Principle of the (hap)FLK tests</h2>
<ul>
<li>FLK and hapFLK are tests aimed at detecting selection based on
population differentiation</li>
<li>The principle is:
<ol>
<li><b>Model  differences in allele frequencies</b> betwen a
group of populations evolving under drift.</li>
<li>At a given locus: <b>Are observed differences with this model ?</b> = test for selection at this locus.</li>
<li>Single SNP test: FLK / LD-based test : hapFLK</li>

</ol></li>

</ul>

</section>
<section id="slide-org0c3b46a">
<h3 id="org0c3b46a">Neutral model in a single population</h3>
<p>
Consider the trajectory of an allele through time in a population of
finite size 
</p>

<div class="figure">
<p><img src="practical/ptraj.png" alt="ptraj.png" />
</p>
</div>

</section>
<section >

<p>
If we were to look at many loci starting from the same initial
frequency \(p_0\). 
</p>
<div class="column" style="float:left; width: 40%">

<div class="figure">
<p><img src="practical/ptrajmulti.png" alt="ptrajmulti.png" />
</p>
</div>
</div>

<div class="column" style="float:right; width: 60%">

<p>
\[E(p) = p_0\]
</p>

<p>
The variance of the final frequencies depends on:
</p>
<ul>
<li>The number of generations (t)</li>
<li>The population size (N)</li>
<li>The initial allele frequency (\(p_0\))</li>

</ul>

<p>
\[Var(p) \approx \frac{t}{2N} p_0(1-p_0)\]
</p>
</div>

</section>
<section id="slide-org342ef42">
<h3 id="org342ef42">Neutral model in multiple populations</h3>
<div class="column" style="float:right; width: 60%">
<p>
If we had a set of loci with identical \(p_0\):
</p>
<ul>
<li>we could estimate \(F \approx \frac{t}{2N}\) and</li>
<li>characterize the distribution of allele frequencies under Neutral
evolution.</li>

</ul>


<p>
<b>Considering the same locus in a set of populations achieves this.</b>
</p>

</div>

<div class="column" style="float:right; width: 40%">

<div class="figure">
<p><img src="practical/startree.png" alt="startree.png" />
</p>
</div>

</div>

<p>
\(F\) characterizes the amount of drift since ancestral split:
shared across (neutral) loci:  estimated by considering all loci.
</p>

</section>
<section id="slide-org9bfc66c">
<h4 id="org9bfc66c">A more general scenario</h4>
<ul>
<li>Populations have different sizes</li>
<li>Successive splits: some populations share ancestry <b>after</b> the
ancestral population.</li>

</ul>

<div class="column" style="float:left; width: 50%">

<div class="figure">
<p><img src="practical/tree.png" alt="tree.png" />
</p>
</div>

</div>

<div class="column" style="float:right; width: 50%">

<div class="figure">
<p><img src="practical/covptraj.png" alt="covptraj.png" />
</p>
</div>

</div>

</section>
<section id="slide-org9caba59">
<h4 id="org9caba59">Exploiting Linkage Disequilibrium: hapFLK</h4>
<div class="column" style="float:left; width: 50%">


<div class="figure">
<p><img src="practical/fastphase.png" alt="fastphase.png" /> 
</p>
</div>
</div>
<div class="column" style="float:right; width: 50%">
<ul>
<li>hapFLK uses the  <a href="https://www.ncbi.nlm.nih.gov/pubmed/16532393">Scheet and Stephens LD model</a> .</li>
<li>Models local similarity between haplotypes via reduction of
dimension: <b>local haplotype clusters</b></li>
<li>Definition of clusters change along the chromosome: accounts for the
effects of recombination</li>

</ul>
</div>

</section>
<section id="slide-org2ee1c07">
<h4 id="org2ee1c07">the hapFLK test</h4>
<ul>
<li>Principle:
<ul>
<li>Consider clusters as alleles</li>
<li>Estimate haplotype cluster frequencies</li>
<li>Test if differences in allele frequencies fit a neutral model.</li>

</ul></li>
<li>Advantages of using this LD model:
<ul>
<li>No need for sliding windows</li>
<li>Can be estimated on unphased genotype data</li>
<li>Admits missing data</li>

</ul></li>
<li><b>The null distribution of hapFLK is not known but can be approximated from the data</b></li>

</ul>

</section>
</section>
<section>
<section id="slide-org98cd3d8">
<h2 id="org98cd3d8">Software overview</h2>
<ul>
<li><code>hapflk</code> is a python program that is meant to be used on the command
line.</li>
<li>Input files should be in <code>plink</code> format. Preferably binary
<code>{bed,bim,fam}</code> files.</li>
<li>First column (FID column) must indicate the population of origin
of an individual</li>

</ul>

</section>
</section>
<section>
<section id="slide-org443fe15">
<h2 id="org443fe15">Practical</h2>
<p>
For this practical we will analyse data from
<a href="http://www.sheephapmap.org/hapmap.php">The SheepHapMap project</a>. First download the data:
</p>
<div class="org-src-container">

<pre><code class="sh" >mkdir -p practical/data/
## get input files into practical directory 
wget -P practical/data http://genoweb.toulouse.inra.fr/~servin/data/NEU.fam 
wget -P practical/data http://genoweb.toulouse.inra.fr/~servin/data/NEU.bim
wget -P practical/data http://genoweb.toulouse.inra.fr/~servin/data/NEU.bed
</code></pre>
</div>

<ul>
<li>These data consists in 7 populations from Northern Europe.</li>
<li>One of them, the Soay, will be considered as an outgroup.</li>

</ul>
</section>
<section id="slide-org96fcf30">
<h3 id="org96fcf30">Estimating population history from SNP data calculating FLK</h3>
<p>
The first task will be aimed at running hapflk in SNP mode, to:
</p>
<ul>
<li>calculate allele frequency</li>
<li>estimate genetic distances between populations (Fsts)</li>
<li>build a population tree</li>
<li>Calculate FLK tests</li>

</ul>
<p>
This is done by running:
</p>
<div class="org-src-container">

<pre><code class="sh" >hapflk --bfile practical/data/NEU --outgroup=Soay -p practical/NEU
</code></pre>
</div>


<p>
Which should take ~ 2 minutes.
</p>

</section>
<section >
<p>
This leads to the following output files:
</p>
<div class="org-src-container">

<pre><code class="sh" >ls practical/NEU*
</code></pre>
</div>

<pre class="example">
practical/NEU_fij.txt
practical/NEU.flk
practical/NEU.frq
practical/NEU_reynolds.txt
practical/NEU_tree.txt
</pre>

<p>
Let&rsquo;s look at  the population tree (<code>NEU_tree.txt</code>). 
</p>
<div class="column" style="float:left; width: 70%">
<div class="org-src-container">

<pre><code class="R" >library(ape)
neu.t=read.tree('practical/NEU_tree.txt')
plot(neu.t,align=T)
axis(1,line=1.5)
title(xlab='F')
</code></pre>
</div>
<p>
Notice that the outgroup (Soay) has been removed.
</p>
</div>

<div class="column" style="float:right; width: 30%">

<div class="figure">
<p><img src="practical/kinship.png" alt="kinship.png" />
</p>
</div>
</div>

</section>
<section >
<p>
The kinship file contains the (co)variance matrix of allele
frequencies estimated with all SNPs.
</p>
<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />
</colgroup>
<tbody>
<tr>
<td class="org-left">ScottishTexel</td>
<td class="org-right">0.175</td>
<td class="org-right">0.091</td>
<td class="org-right">0.085</td>
<td class="org-right">0.037</td>
<td class="org-right">0.015</td>
<td class="org-right">0</td>
</tr>

<tr>
<td class="org-left">NewZealandTexel</td>
<td class="org-right">0.091</td>
<td class="org-right">0.161</td>
<td class="org-right">0.085</td>
<td class="org-right">0.037</td>
<td class="org-right">0.015</td>
<td class="org-right">0</td>
</tr>

<tr>
<td class="org-left">GermanTexel</td>
<td class="org-right">0.085</td>
<td class="org-right">0.085</td>
<td class="org-right">0.119</td>
<td class="org-right">0.037</td>
<td class="org-right">0.015</td>
<td class="org-right">0</td>
</tr>

<tr>
<td class="org-left">Galway</td>
<td class="org-right">0.037</td>
<td class="org-right">0.037</td>
<td class="org-right">0.037</td>
<td class="org-right">0.155</td>
<td class="org-right">0.015</td>
<td class="org-right">0</td>
</tr>

<tr>
<td class="org-left">NewZealandRomney</td>
<td class="org-right">0.015</td>
<td class="org-right">0.015</td>
<td class="org-right">0.015</td>
<td class="org-right">0.015</td>
<td class="org-right">0.125</td>
<td class="org-right">0</td>
</tr>

<tr>
<td class="org-left">IrishSuffolk</td>
<td class="org-right">0</td>
<td class="org-right">0</td>
<td class="org-right">0</td>
<td class="org-right">0</td>
<td class="org-right">0</td>
<td class="org-right">0.159</td>
</tr>
</tbody>
</table>


</section>
<section >
<p>
The <code>NEU.flk</code> file contains results of the FLK test.
</p>


<p>
We can check the fit of the model by verifying that the empirical
distribution of FLK is close to a \(\chi^2(5)\) 
</p>

<div class="column" style="float:left; width: 70%">
<div class="org-src-container">

<pre><code class="R" >flk=read.table('practical/NEU.flk',h=T)
hist(flk$flk,n=100,f=F)
xx=seq(0,50,0.01)
lines(xx,dchisq(xx,df=5),lwd=2,col=2)
</code></pre>
</div>

</div>

<div class="column" style="float:right; width: 30%">

<div class="figure">
<p><img src="practical/flkdist.png" alt="flkdist.png" />
</p>
</div>
</div>
</section>
<section id="slide-org511bcb5">
<h3 id="org511bcb5">Running hapFLK on a (bit of) chromosome</h3>
<ul>
<li>hapFLK is a test based on LD patterns.</li>
<li>LD patterns are relevant for linked markers</li>
<li>So hapFLK needs to be calculated independently for each chromosome :)</li>
<li>To reduce computation time, we will practice on a small, non random
region of a chromosome</li>
<li>We use plink to create our reduced dataset:</li>

</ul>
<div class="org-src-container">

<pre><code class="sh" >plink --sheep --bfile practical/data/NEU \
      --chr 2 --from-kb 100000 --to-kb 140000\
      --out practical/data/mstn --make-bed
</code></pre>
</div>

</section>
<section id="slide-orgbe22cde">
<h4 id="orgbe22cde">Running the LD model</h4>
<p>
As for FLK, hapFLK must be calculated using a kinship matrix
<b>estimated genome-wide</b>. 
</p>

<p>
When using the LD model, the kinship <b>must</b> be provided.
</p>

<p>
hapFLK calculations are turned on by setting a number of haplotype
clusters using the <code>-K</code> flag.
</p>

</section>
<section >

<div class="org-src-container">

<pre><code class="sh" >hapflk --bfile practical/data/mstn \
       --outgroup=Soay \
       -p practical/mstn \
       --annot \
       --kinship practical/NEU_fij.txt \
       -K 15 \
       --nfit=2 \
       --ncpu=2
</code></pre>
</div>


<ul>
<li><b>-K 15</b>: the number of clusters to model LD. This <b>depends on the
data</b>. Can be estimated using <a href="http://scheet.org/software.html"><code>fastPHASE</code></a> cross-validation
proceudre.</li>
<li><b>&#x2013;annot</b>: turn on the production of output files  to
annotate the FLK and hapFLK signal. Use when looking at small
genomic regions (some files can get big !).</li>
<li><b>&#x2013;nfit=2</b>: this is only used here to reduce computation
time. <b>Don&rsquo;t do this at home</b> (keep the default unless you know what
you are doing).</li>
<li><b>&#x2013;ncpu=2</b>: use more if you can.</li>

</ul>

</section>
<section id="slide-orgac16d7b">
<h4 id="orgac16d7b">hapFLK output files</h4>
<div class="column" style="float:left; width: 50%">
<div class="org-src-container">

<pre><code class="sh" >ls practical/mstn*
</code></pre>
</div>

<pre class="example">
practical/mstn.eig
practical/mstn.flk
practical/mstn.flk.eig
practical/mstn.frq
practical/mstn.hapflk
practical/mstn.hapflk.eig
practical/mstn.kfrq.fit_0.bz2
practical/mstn.kfrq.fit_1.bz2
practical/mstn.png
practical/mstn.rey
</pre>
</div>
<div class="column" style="float:left; width: 50%">
<ul>
<li><code>mstn.flk</code> and <code>mstn.frq</code>: same as before. Sometimes useful (not here).</li>
<li><code>mstn.hapflk</code>: contains the hapFLK results</li>
<li><code>mstn.kfrq.fit_{N}.bz2</code>: frequency of haplotype clusters in
populations</li>
<li><code>mstn.rey</code>: estimates of local \(F_{_ST}\) between populations</li>
<li><code>mstn.*.eig</code>: signal decomposition into principal components.</li>

</ul>
</div>

</section>
<section id="slide-org9137d52">
<h4 id="org9137d52">FLK and hapFLK</h4>
<div class="org-src-container">

<pre><code class="R" >flk=read.table('practical/mstn.flk',h=T)
hflk=read.table('practical/mstn.hapflk',h=T)
par(mfrow=c(1,2))
plot(flk$pos,-log10(flk$pvalue), main='FLK',
     xlab='Position', ylab='-log(p-value)', pch=16)
plot(hflk$pos,hflk$hapflk,main='hapFLK', type='l', lwd=2,
     xlab='Position', ylab='hapFLK value')
</code></pre>
</div>


<div class="figure">
<p><img src="practical/mstn.png" alt="mstn.png" />
</p>
</div>

</section>
<section >

<div class="figure">
<p><img src="practical/mstn.png" alt="mstn.png" />
</p>
</div>

<p>
NB: hapFLK results are not p-values. These can be computed <b>after a
whole genome scan</b> using the <a href="https://forge-dga.jouy.inra.fr/attachments/download/5765/scaling_chi2_hapflk.py"><code>scaling_chi2_hapflk.py</code></a> script available 
on the hapFLK website. 
</p>

</section>
<section id="slide-orgcb1ad72">
<h4 id="orgcb1ad72">Local cluster plots</h4>
<p>
The <a href="https://forge-dga.jouy.inra.fr/attachments/download/2919/hapflk-clusterplot.R">hapflk-clusterplot.R</a> script  produces local cluster plots.
</p>
<div class="org-src-container">

<pre><code class="sh" >## chmod +x hapflk-clusterplot.R
hapflk-clusterplot.R practical/mstn.kfrq.fit_0.bz2
</code></pre>
</div>


<div class="figure">
<p><img src=" practical/clusterplot.png" alt="clusterplot.png" width="40%" />
</p>
</div>

</section>
<section id="slide-org2d92725">
<h4 id="org2d92725">Local population trees</h4>
<p>
Keep the tree structure but reesimates branch length. Use
<a href="https://forge-dga.jouy.inra.fr/attachments/download/3248/local_reynolds.py"><code>local_reynolds.py</code></a> and <a href="https://forge-dga.jouy.inra.fr/attachments/download/3272/local_trees.R"><code>local_trees.R</code></a>
</p>

<div class="org-src-container">

<pre><code class="sh" >## chmod +x local_reynolds.py
cd practical/
local_reynolds.py -p mstn
## produces two files
## hapflk_snp_reynolds.txt
## hapflk_hap_reynolds.txt

## EDIT the local_trees.R script (available on github)
cd ../../
R CMD BATCH local_trees.R
</code></pre>
</div>

</section>
<section >

<div class="figure">
<p><img src="practical/localSnpTree.png" alt="localSnpTree.png" width="50%" />
</p>
</div>
</section>
</section>
</div>
</div>
<script src="http://cdn.jsdelivr.net/reveal.js/3.0.0/lib/js/head.min.js"></script>
<script src="http://cdn.jsdelivr.net/reveal.js/3.0.0/js/reveal.js"></script>

<script>
// Full list of configuration options available here:
// https://github.com/hakimel/reveal.js#configuration
Reveal.initialize({

controls: true,
progress: true,
history: false,
center: false,
slideNumber: 'c',
rollingLinks: false,
keyboard: true,
overview: true,

theme: Reveal.getQueryHash().theme, // available themes are in /css/theme
transition: Reveal.getQueryHash().transition || 'none', // default/cube/page/concave/zoom/linear/fade/none
transitionSpeed: 'default',
multiplex: {
    secret: '', // null if client
    id: '', // id, obtained from socket.io server
    url: '' // Location of socket.io server
},

// Optional libraries used to extend on reveal.js
dependencies: [
 { src: 'http://cdn.jsdelivr.net/reveal.js/3.0.0/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } }]
});
</script>
</body>
</html>
